File: Práctica4/lib/PowerEfficientSpaceStation.rb
#encoding:utf-8

require_relative 'SpaceStation'
require_relative 'Transformation'
require_relative 'PowerEfficientSpaceStationToUI'
#require_relative 'SuppliesPackage'

module Deepspace
  # Clase que representa una estación espacial eficiente
  # Una estación espacial eficiente recibe una bonificación de potencia de disparo y escudo
  class PowerEfficientSpaceStation < SpaceStation
    # @!attribute [Float] factor de eficiencia
    @@EFFICIENCYFACTOR = 1.10

    # Constructor
    # @param station [SpaceStation] estación espacial a convertir en estación eficiente
    # --Overriden
    def initialize(station)
      copy(station)
    end

    # Realiza un disparo
    # @return [Float] potencia disparo
    # --Overriden
    def fire
      super * @@EFFICIENCYFACTOR
    end

    # Usa el escudo de protección
    # @return [Float] potencia del escudo
    # --Overriden
    def protection
      super * @@EFFICIENCYFACTOR
    end

    # Recibe un loot
    # @param [Loot] loot a recibir
    # @return [Transformation] si la transformacion ha sido a eficiente (GETEFFICIENT)
    # en caso contrario NOTRANSFORM
    def setLoot(loot)
      super
      if loot.efficient
        Transformation::GETEFFICIENT
      else
        Transformation::NOTRANSFORM
      end
    end
  end
end

File: Práctica5/java/src/deepspace/PowerEfficientSpaceStation.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package deepspace;

/**
 * Clase que representa una estación espacial eficiente.
 * Una estación espacial eficiente consigue mayor potencia de disparo y de escudp
 */
public class PowerEfficientSpaceStation extends SpaceStation{
    /**
     * Factor de eficiencia que gana la nave
     */
    private static final float EFFICIENCYFACTOR = 1.1f;
    
    /**
     * Constructor
     * @param station estación a convertir en eficiente
     */
    public PowerEfficientSpaceStation(SpaceStation station) {
        super(station);
    }
    
    /**
     * Realiza un disparo
     * @return potencia de disparo
     */
    @Override
    public float fire(){
        return super.fire()*EFFICIENCYFACTOR;
    }
    
    /**
     * Utiliza el escudo
     * @return potencia del escudp
     */
    @Override
    public float protection(){
        return super.protection()*EFFICIENCYFACTOR;
    }
    
    /**
     * Recibe y procesa un loot
     * @param loot loot a procesar
     * @return transformación que sufre la nave
     */
    @Override
    public Transformation setLoot(Loot loot){
        Transformation trans = super.setLoot(loot);
        if (trans == Transformation.SPACECITY)
            return Transformation.NOTRANSFORM;
        else
            return trans;
    }
    
    /**
     * To UI.
     * @return 
     */
    @Override
    public PowerEfficientSpaceStationToUI getUIversion() {
        return new PowerEfficientSpaceStationToUI(this);
    }
}
